exp/addons/internal/controllers/clusterresourceset_controller.go

```diff
@@ -131,17 +131,18 @@ func (r *ClusterResourceSetReconciler) Reconcile(ctx context.Context, req ctrl.R
 		return ctrl.Result{}, err
 	}
 
-	// Add finalizer first if not exist to avoid the race condition between init and delete
-	if !controllerutil.ContainsFinalizer(clusterResourceSet, addonsv1.ClusterResourceSetFinalizer) {
-		controllerutil.AddFinalizer(clusterResourceSet, addonsv1.ClusterResourceSetFinalizer)
-		return ctrl.Result{}, nil
-	}
-
 	// Handle deletion reconciliation loop.
 	if !clusterResourceSet.ObjectMeta.DeletionTimestamp.IsZero() {
 		return r.reconcileDelete(ctx, clusters, clusterResourceSet)
 	}
 
+	// Add finalizer first if not set to avoid the race condition between init and delete.
+	// Note: Finalizers in general can only be added when the deletionTimestamp is not set.
+	if !controllerutil.ContainsFinalizer(clusterResourceSet, addonsv1.ClusterResourceSetFinalizer) {
```

Comment:
I'm not sure this is correct.
Without a finalizer delete will leave orphan objects (or at least this is what I remember from issues that forced us to add the finalizers as a first thing in all controllers)

Reply:
The apiserver will simply not allow us to add a finalizer once the deletionTimestamp is set.

> "Reconciler error" err="Cluster.cluster.x-k8s.io "sgcluster" is invalid: metadata.finalizers: Forbidden: no new finalizers can be added if the object is being deleted, found new finalizers []string{"cluster.cluster.x-k8s.io"}"

In my opinion it is important that we add the finalizer synchronously before we go into "reconcile normal". It is not possible to set the finalizer when the deletionTimestamp is set before "reconcile delete".

I think it's enough to only do this before reconcile normal as this is where we actualy create new objects/infra/....

Reply:
Note we still do it as the "first thing" before reconcileNormal. Just not as the first thing before reconcileDelete anymore (because adding a finalizer in that case will be rejected by the apiserver)

Reply:
Ack, thanks for clarification